# Data analysis of TEQUILA-seq

## About

This workflow is for analyzing the data generated by TEQUILA-seq (Target Enrichment and Quantification Utilizing Isothermally Linear-Amplified probes in conjunction with long-read Sequencing). <br/><br/>
The workflow consists of two parts: 1) Identification and quantification of isoform using ESPRESSO tool (this pipeline has been wrapped up as a snakemake workflow) and 2) downstream data analysis and visualization. 


## Table of contents

* [Data process](#data-process)
  + [Install](#install)
  + [Usage](#usage)
  + [Configuration](#configuration)
* [Data analysis and visualization](#data-analysis-and-visualization)


## Data process

#### Install

[Conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/linux.html) must be installed first. Then the detailed steps are as follows:

1. Enter `TEQUILA-seq` folder.
2. Modify `CONDA_ENV_PREFIX` paths in [set_env_vars.sh](set_env_vars.sh)
3. Install other dependencies using conda: `./install`.
4. Manually install following packages:
    + argparse: `conda install -c conda-forge configargparse`
5. Download reference genome sequence file and put it into the 'references' folder:         https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/GRCh37_mapping/gencode.v34lift37.annotation.gtf.gz
6. Modify some absolute file paths in [snakemake_config.yaml](snakemake_config.yaml).
    + In this project, we mainly have three sample groups: 'brain sample + SIRVset4', 'SH-SY5Y cell + SIRVset4' and 'BRCA cell lines'.
    + Configure files of all sample groups are pre-defined in the `Config_for_each_system`. When performing the analysis, the desired `snakemake_config.yaml` should be copied into main folder.
    + The `visualization_path` and `conda_wrapper` of `snakemake_config.yaml` should be modified accordingly.

#### Usage

Note: all commands should be run under the conda environment: 
1. Modify `snakemake_config.yaml` accordingly.
2. Submit the job with `sbatch --time=7-0:0:0 ./run` (Base-calling, alignment and isoform identification based on all samples)


#### Configuration

[snakemake_config.yaml](snakemake_config.yaml)
* `long_read_samples`
  + each sample has its own entry which defines the `sample_name`
  + if starting from fast5 files then under `sample_name` set: `fast5_dir` and `guppy_config`: e.g.
```
    samples:
      RNA_sample_1:
        - guppy_config: 'dna_r9.4.1_450bps_fast.cfg'
          fast5_dir: '~/fast5/'
```
  + if starting from a fastq file then under `sample_name` set: `fastq`: e.g.
```
    samples:
      RNA_sample_1:
        - fastq: '~/combined.fastq'
```
* `guppy_bin_path: '/path/to/guppy/bin/'` (only needed if starting from fast5 files)
* Specify the reference files to use:
  + Either manually put each file in `references/` or provide a url to download the (potentially gzipped) file:
  + `gtf_name: 'file_name.gtf'`
  + `gff3_name: 'file_name.gff3'`
  + `fasta_name: 'file_name.fasta'`
* `conda_wrapper:`: based on current folder
* `espresso_path:`: based on current folder
```
reference_files:
  file_name.gtf.gz:
    url: 'protocol://url/for/file_name.gtf.gz'
```


## Data analysis and visualization

### IMPACT genes panel on BRCA cell lines 
Note: all commands should be run under the conda environment: 
1. Manually install following packages:
  + R: `conda config --add channels conda-forge`
  + `conda config --set channel_priority strict`
  + `conda create -n r_plot r-essentials r-base=4.0.5`
  + `conda activate r_plot`
  + ggplot2: `conda install -c r r-ggplot2`
  + tidyverse: `conda install -c r r-tidyverse`
  + ggplotify: `conda install -c conda-forge r-ggplotify`
  + numpy for python: `conda install -c anaconda numpy`
  + scales: `conda install -c r r-scales`
  + forcats: `conda install -c conda-forge r-forcats`
  
2. Generate figures for selected examples.
  + Enter [Examples_visualization](Examples_visualization) folder.
  + Run `sh  Examples_visualization.sh`.
  + Check figures generated in [Example_res](./Examples_visualization/Example_res) folder.
  
